### PR TITLE
HttpOverrides support for WebSocket

### PR DESCRIPTION
**Pull request by tauu**
This PR fixes WebSocket.connect not using any HttpClient provided by HttpOverrides. Currently the HttpClient used for WebSocket connections is initialised in a static field. Consequently any HttpClient set in HttpOverrides after the first initialisation is not taken into account for WebSocket connections. This is changed by initialising a new HttpClient for each WebSocket connection.

As it is currently possible to set the user agent for the HttpClient used for the websocket request, the current behaviour is retained:
  - If a custom HttpClient is provided in the connect call, the userAgent field is not used.
  - If no custom HttpClient is provided in the connect call and the userAgent field was not set, the userAgent of the default HttpClient is not set.
  - If no custom HttpClient is provided in the connect call and the userAgent field was set, the userAgent of the default HttpClient is set.
This is accomplished by adding the private field _overrideUserAgent and setting it the first time the userAgent field ist set. This extra field is necessary, as null is a valid value for the userAgent property of a HttpClient and it is not possible to distinguish between _userAgent being initialised to null or it being set to null otherwise.

Two tests are included which provide two different overrides and verify that these lead to results for the same WebSocket.connect call.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

### CODE REVIEW

**Comment by brianquinlan** [comment]
I left some feedback here:
https://dart-review.googlesource.com/c/sdk/+/305620/2/sdk/lib/_http/websocket_impl.dart#b992

This is a potentially breaking change which needs to follow the [Dart SDK breaking change process](https://github.com/dart-lang/sdk/blob/master/docs/process/breaking-changes.md).

Would you like for me to deal with that bureaucracy?

**Comment by tauu** [comment]
First of all, sorry for the slow response. I wrongly assumed, that I would get a notification from google source / copybara would relay comments to github. But apparently that did not happen 😅 and I was on vacation once the github comment appeared. Thank you for the review. I will address the comments in dart-review.

> Would you like for me to deal with that bureaucracy?
That would be amazing!!

**Comment by mraleph** [comment]
@brianquinlan are you taking over this? Should this PR be closed?

**Comment by brianquinlan** [comment]
@tauu 

I have a proposed set of simplified tests: https://dart-review.googlesource.com/c/sdk/+/305620/7/tests/standalone/io/http_override_websocket_test.dart

Let me know what you think.

File: sdk/lib/_http/websocket_impl.dart [hunk 1]
```diff
@@ -1263,11 +1269,12 @@ class _WebSocketImpl extends Stream with _ServiceObject implements WebSocket {
     }
     return _sink.close();
   }
-
-  static String? get userAgent => _httpClient.userAgent;
+
```
Comment (albertms10) [inline sdk/lib/_http/websocket_impl.dart:1272]: Unwanted whitespace? 👀

**Comment by brianquinlan** [comment]
Unfortunately, I don't think that this change is viable because it will break existing code. For example, `flutter test` uses `HttpOverrides` to replace `HttpClient` with a mock but using that mock in `WebSocket` will take away the ability to talk to VM services. See https://github.com/dart-lang/sdk/commit/1b8382b14d5cf24d6d5ebe3a2d74fba286c45a12

**Comment by mraleph** [comment]
@brianquinlan could not we just fix the tests by undoing the override? These tests starting to fail is in spirit of the change / flutter test - we have discovered tests which are using actual network connection instead of a mocked connection. Test should be explicit about that.

**Comment by brianquinlan** [comment]
@mraleph I'm not sure that I understand your question.

[`flutter_test` defines its own `HttpOverrides`](https://github.com/flutter/flutter/blob/0dc066b717a459f3544b2b074ffc35ba7c351e30/packages/flutter_test/lib/src/_binding_io.dart#L71) that disables all HTTP requests and [installs it globally](https://github.com/flutter/flutter/blob/0dc066b717a459f3544b2b074ffc35ba7c351e30/packages/flutter_test/lib/src/_binding_io.dart#L27)

 The tests were failing because the test infrastructure itself uses `WebSocket`, which will end up using the `_MockHttpClient`.

I can probably fix `flutter_test` but I'm worried about other code out there that consumes VM services and will be broken by this change.

**Pull request closed without merging**